File: android/feature/client/src/main/kotlin/com/devora/mencare/feature/client/profile/ProfileScreen.kt
package com.devora.mencare.feature.client.profile

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.Logout
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.devora.mencare.core.common.formatDayGroup
import com.devora.mencare.core.common.formatMonthYear
import com.devora.mencare.core.common.formatTime
import com.devora.mencare.core.designsystem.component.AccentButton
import com.devora.mencare.core.designsystem.component.BarAction
import com.devora.mencare.core.designsystem.component.BrandSectionLabel
import com.devora.mencare.core.designsystem.component.BrandSwitch
import com.devora.mencare.core.designsystem.component.DarkHeader
import com.devora.mencare.core.designsystem.component.EmailField
import com.devora.mencare.core.designsystem.component.InlineErrorBanner
import com.devora.mencare.core.designsystem.component.NameField
import com.devora.mencare.core.designsystem.component.PhoneField
import com.devora.mencare.core.designsystem.component.StoneKeyValueRow
import com.devora.mencare.core.designsystem.component.readableWidth
import com.devora.mencare.core.designsystem.theme.Bone
import com.devora.mencare.core.designsystem.theme.Cormorant
import com.devora.mencare.core.designsystem.theme.ErrorRed
import com.devora.mencare.core.designsystem.theme.Ink
import com.devora.mencare.core.designsystem.theme.OliveWood
import com.devora.mencare.core.designsystem.theme.Stone
import com.devora.mencare.core.designsystem.theme.StoneBorder
import com.devora.mencare.core.designsystem.theme.TextMuted
import com.devora.mencare.core.model.groupConsecutiveDays
import com.devora.mencare.feature.client.R

@Composable
fun ProfileScreen(
    onLoggedOut: () -> Unit,
    viewModel: ProfileViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    var editing by rememberSaveable { mutableStateOf(false) }
    var changingPassword by rememberSaveable { mutableStateOf(false) }

    LaunchedEffect(state.loggedOut) {
        if (state.loggedOut) onLoggedOut()
    }

    // Si esce dalla modifica solo quando il salvataggio è andato a buon fine.
    LaunchedEffect(state.saved) {
        if (state.saved) editing = false
    }

    Column(modifier = Modifier.fillMaxSize().background(Bone)) {
        DarkHeader(
            contentPadding = PaddingValues(start = 20.dp, end = 20.dp, top = 12.dp, bottom = 20.dp),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Box(
                    modifier = Modifier
                        .size(56.dp)
                        .clip(RoundedCornerShape(16.dp))
                        .background(OliveWood),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        state.user?.initials.orEmpty().uppercase(),
                        fontFamily = Cormorant,
                        fontSize = 19.sp,
                        color = Bone,
                    )
                }
                Column(Modifier.weight(1f).padding(start = 14.dp)) {
                    Text(
                        state.user?.fullName.orEmpty(),
                        style = MaterialTheme.typography.headlineMedium.copy(fontSize = 26.sp),
                        color = Bone,
                    )
                    Text(
                        stringResource(
                            R.string.profile_since,
                            state.user?.memberSince?.let { formatMonthYear(it) }.orEmpty(),
                            state.user?.visitCount ?: 0,
                        ),
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                        color = Bone,
                    )
                }
                BarAction(
                    text = stringResource(if (editing) R.string.profile_save else R.string.profile_edit),
                    onClick = { if (editing) viewModel.save() else editing = true },
                    color = OliveWood,
                )
            }
        }

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .readableWidth()
                .imePadding()
                .padding(horizontal = 20.dp)
                .padding(top = 18.dp, bottom = 24.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            // Il profilo ha sempre qualcosa da mostrare (i dati della sessione),
            // quindi un guasto di lettura è una riga e non una schermata intera.
            state.loadError?.let { failure ->
                InlineErrorBanner(error = failure, onRetry = viewModel::retry)
            }
            BrandSectionLabel(stringResource(R.string.profile_personal_data))
            if (editing) {
                NameField(
                    value = state.firstName,
                    onValueChange = { viewModel.onFieldChange(ProfileField.FIRST_NAME, it) },
                    label = stringResource(R.string.profile_first_name),
                    error = state.fieldErrors[ProfileField.FIRST_NAME],
                )
                NameField(
                    value = state.lastName,
                    onValueChange = { viewModel.onFieldChange(ProfileField.LAST_NAME, it) },
                    label = stringResource(R.string.profile_last_name),
                    error = state.fieldErrors[ProfileField.LAST_NAME],
                )
                EmailField(
                    value = state.email,
                    onValueChange = { viewModel.onFieldChange(ProfileField.EMAIL, it) },
                    label = stringResource(R.string.profile_email),
                    error = state.fieldErrors[ProfileField.EMAIL],
                )
                PhoneField(
                    value = state.phone,
                    onValueChange = { viewModel.onFieldChange(ProfileField.PHONE, it) },
                    label = stringResource(R.string.profile_phone),
                    error = state.fieldErrors[ProfileField.PHONE],
                )
                if (state.saveFailed) {
                    Text(
                        stringResource(R.string.profile_error_save),
                        style = MaterialTheme.typography.bodySmall,
                        color = ErrorRed,
                    )
                }
            } else {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(16.dp))
                        .background(Stone),
                ) {
                    StoneKeyValueRow(
                        stringResource(R.string.profile_first_name),
                        "${state.firstName} ${state.lastName}".trim(),
                    )
                    HorizontalDivider(color = StoneBorder)
                    StoneKeyValueRow(stringResource(R.string.profile_email), state.email)
                    HorizontalDivider(color = StoneBorder)
                    StoneKeyValueRow(stringResource(R.string.profile_phone), state.phone)
                }
            }

            // The salon's own opening hours, as the owner sets them in Gestione.
            state.salon?.let { salon ->
                val groups = salon.weeklyHours.groupConsecutiveDays()
                if (groups.any { it.second != null }) {
                    Spacer(Modifier.height(6.dp))
                    BrandSectionLabel(stringResource(R.string.profile_salon_hours))
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        groups.forEach { (days, range) ->
                            val closed = range == null
                            Box(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .clip(RoundedCornerShape(10.dp))
                                    .then(
                                        if (closed) {
                                            Modifier.border(1.dp, StoneBorder, RoundedCornerShape(10.dp))
                                        } else {
                                            Modifier.background(Stone)
                                        },
                                    ),
                            ) {
                                StoneKeyValueRow(
                                    label = formatDayGroup(days),
                                    value = range?.let { "${formatTime(it.start)} — ${formatTime(it.end)}" }
                                        ?: stringResource(R.string.profile_salon_closed),
                                    valueColor = if (closed) TextMuted else Ink,
                                )
                            }
                        }
                    }
                    Text(
                        salon.address,
                        style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                        color = TextMuted,
                        modifier = Modifier.padding(start = 4.dp, top = 2.dp),
                    )
                }
            }

            Spacer(Modifier.height(6.dp))
            BrandSectionLabel(stringResource(R.string.profile_account_management))
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(16.dp))
                    .background(Stone),
            ) {
                StoneKeyValueRow(
                    label = stringResource(R.string.profile_change_password),
                    value = "••••••••",
                    valueColor = OliveWood,
                    onClick = { changingPassword = true },
                )
            }

            Spacer(Modifier.height(6.dp))
            BrandSectionLabel(stringResource(R.string.profile_notifications))
            PrefRow(
                title = stringResource(R.string.profile_pref_reminder),
                hint = stringResource(R.string.profile_pref_reminder_hint),
                checked = state.prefs.appointmentReminder,
                onCheckedChange = { viewModel.onPrefChange(state.prefs.copy(appointmentReminder = it)) },
            )
            PrefRow(
                title = stringResource(R.string.profile_pref_waitlist),
                hint = null,
                checked = state.prefs.waitlistAlerts,
                onCheckedChange = { viewModel.onPrefChange(state.prefs.copy(waitlistAlerts = it)) },
            )
            PrefRow(
                title = stringResource(R.string.profile_pref_marketing),
                hint = null,
                checked = state.prefs.marketing,
                onCheckedChange = { viewModel.onPrefChange(state.prefs.copy(marketing = it)) },
            )

            Spacer(Modifier.height(10.dp))
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(56.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .border(1.dp, StoneBorder, RoundedCornerShape(16.dp))
                    .clickable(onClick = viewModel::logout),
                horizontalArrangement = Arrangement.Center,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    Icons.AutoMirrored.Outlined.Logout,
                    contentDescription = null,
                    tint = ErrorRed,
                    modifier = Modifier.size(18.dp),
                )
                Text(
                    stringResource(R.string.profile_logout),
                    style = MaterialTheme.typography.titleMedium,
                    color = Ink,
                    modifier = Modifier.padding(start = 10.dp),
                )
            }

            if (editing && state.dirty) {
                AccentButton(
                    text = stringResource(R.string.profile_save),
                    onClick = viewModel::save,
                    loading = state.saving,
                    height = 54.dp,
                    shape = RoundedCornerShape(16.dp),
                )
            }
        }
    }

    if (changingPassword) {
        ChangePasswordScreen(onDismiss = { changingPassword = false })
    }
}

@Composable
private fun PrefRow(
    title: String,
    hint: String?,
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(Stone)
            .padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(title, style = MaterialTheme.typography.bodyLarge, color = Ink)
            if (hint != null) {
                Text(
                    hint,
                    style = MaterialTheme.typography.bodySmall.copy(fontSize = 12.sp),
                    color = TextMuted,
                )
            }
        }
        BrandSwitch(checked = checked, onCheckedChange = onCheckedChange)
    }
}

File: ios/MenCare/Features/Client/Profile/ProfileScreen.swift
import SwiftUI
import Observation

@MainActor
@Observable
final class ProfileViewModel {

    private let auth: AuthRepository
    private let catalog: CatalogRepository

    var firstName = ""
    var lastName = ""
    var email = ""
    var phone = ""
    var dirty = false
    var saved = false
    var saving = false
    /// Chiave del campo → messaggio: stesse regole della registrazione, che qui
    /// prima non c'erano affatto.
    var fieldErrors: [String: String] = [:]
    /// Esito vero del salvataggio: prima `saved` diventava true comunque, anche
    /// quando la chiamata falliva.
    var saveError: String?

    let state = LoadState()
    private(set) var user: User?
    private(set) var salon = Salon(name: "", address: "", city: "")
    private(set) var prefs = ClientNotificationPrefs()
    /// Gli interruttori scrivono subito e non hanno un tasto "Salva": se il
    /// server rifiuta va detto li'.
    var prefsError: String?

    @ObservationIgnored private var submitted = false

    init(auth: AuthRepository, catalog: CatalogRepository) {
        self.auth = auth
        self.catalog = catalog
    }

    func load() async {
        await state.run {
            let me = try await auth.currentUser()
            user = me
            // I campi si ricompilano solo se non li si sta modificando: un
            // ricaricamento non deve cancellare quello che si sta scrivendo.
            if !dirty, let me {
                firstName = me.firstName
                lastName = me.lastName
                email = me.email
                phone = me.phone
            }
            salon = try await catalog.catalog().salon
            prefs = try await auth.notificationPrefs()
        }
    }

    func fieldChanged() {
        dirty = true
        saved = false
        saveError = nil
        if submitted { fieldErrors = validate() }
    }

    private func validate() -> [String: String] {
        var errors: [String: String] = [:]
        errors["firstName"] = validateName(firstName).message
        errors["lastName"] = validateName(lastName).message
        errors["email"] = validateEmail(email).message
        errors["phone"] = validatePhone(phone).message
        return errors.compactMapValues { $0 }
    }

    func save() {
        guard !saving else { return }
        submitted = true
        let errors = validate()
        if !errors.isEmpty {
            fieldErrors = errors
            saveError = L("validation_form_invalid")
            return
        }
        fieldErrors = [:]
        saveError = nil
        saving = true
        let phoneToSave = normalizePhone(phone) ?? phone
        Task {
            let result = await auth.updateProfile(
                firstName: firstName.trimmingCharacters(in: .whitespaces),
                lastName: lastName.trimmingCharacters(in: .whitespaces),
                email: normalizeEmail(email),
                phone: phoneToSave
            )
            saving = false
            switch result {
            case .success(let user):
                phone = user.phone
                email = user.email
                dirty = false
                saved = true
            case .failure(let error):
                saved = false
                saveError = error.displayMessage
            }
        }
    }

    func setPrefs(_ updated: ClientNotificationPrefs) {
        let previous = prefs
        // Aggiornamento ottimista: l'interruttore si muove subito e torna
        // indietro se il server dice di no.
        prefs = updated
        prefsError = nil
        Task {
            switch await auth.updateNotificationPrefs(updated) {
            case .success(let stored): prefs = stored
            case .failure(let error):
                prefs = previous
                prefsError = error.displayMessage
            }
        }
    }

    func logout(onLoggedOut: @escaping () -> Void) {
        Task {
            await auth.logout()
            onLoggedOut()
        }
    }
}

struct ProfileScreen: View {
    @Bindable var viewModel: ProfileViewModel
    let onLoggedOut: () -> Void

    @State private var editing = false
    @State private var changingPassword = false
    @Environment(\.container) private var container

    var body: some View {
        VStack(spacing: 0) {
            DarkHeader(contentPadding: EdgeInsets(top: 12, leading: 20, bottom: 20, trailing: 20)) {
                HStack(spacing: 14) {
                    RoundedRectangle(cornerRadius: 16)
                        .fill(Color.oliveWood)
                        .frame(width: 56, height: 56)
                        .overlay(
                            Text((viewModel.user?.initials ?? "").uppercased())
                                .font(Typo.cormorant(19, weight: .regular))
                                .foregroundStyle(Color.bone)
                        )
                    VStack(alignment: .leading, spacing: 0) {
                        Text(viewModel.user?.fullName ?? "")
                            .font(Typo.cormorant(26))
                            .foregroundStyle(Color.bone)
                        Text(L(
                            "profile_since",
                            viewModel.user.map { formatMonthYear($0.memberSince) } ?? "",
                            viewModel.user?.visitCount ?? 0
                        ))
                        .font(Typo.jost(12))
                        .foregroundStyle(Color.bone)
                    }
                    Spacer()
                    BarAction(
                        text: L(editing ? "profile_save" : "profile_edit"),
                        action: {
                            guard editing else {
                                editing = true
                                return
                            }
                            if viewModel.dirty {
                                viewModel.save()
                                // Si esce dalla modifica solo se i campi reggono:
                                // altrimenti l'errore resterebbe invisibile.
                                editing = !viewModel.fieldErrors.isEmpty
                            } else {
                                editing = false
                            }
                        },
                        color: .oliveLight
                    )
                }
            }

            if viewModel.state.isLoading, !viewModel.state.hasLoadedOnce {
                BrandLoading()
                Spacer()
            } else if let error = viewModel.state.error, !viewModel.state.hasLoadedOnce {
                BrandErrorRetry(error: error, retry: { Task { await viewModel.load() } })
                Spacer()
            } else {
            ScrollView {
                VStack(alignment: .leading, spacing: 10) {
                    if let error = viewModel.state.error {
                        InlineErrorBar(message: error.displayMessage, retry: { Task { await viewModel.load() } })
                    }
                    BrandSectionLabel(text: L("profile_personal_data"))
                    if editing {
                        NameField(
                            label: L("profile_first_name"), text: $viewModel.firstName,
                            error: viewModel.fieldErrors["firstName"]
                        )
                        NameField(
                            label: L("profile_last_name"), text: $viewModel.lastName,
                            error: viewModel.fieldErrors["lastName"]
                        )
                        EmailField(
                            label: L("profile_email"), text: $viewModel.email,
                            error: viewModel.fieldErrors["email"]
                        )
                        PhoneField(
                            label: L("profile_phone"), text: $viewModel.phone,
                            error: viewModel.fieldErrors["phone"]
                        )
                        FormErrorBanner(message: viewModel.saveError)
                    } else {
                        VStack(spacing: 0) {
                            StoneKeyValueRow(label: L("profile_first_name"), value: "\(viewModel.firstName) \(viewModel.lastName)")
                            Rectangle().fill(Color.stoneBorder).frame(height: 1)
                            StoneKeyValueRow(label: L("profile_email"), value: viewModel.email)
                            Rectangle().fill(Color.stoneBorder).frame(height: 1)
                            StoneKeyValueRow(label: L("profile_phone"), value: viewModel.phone)
                        }
                        .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
                        .clipShape(RoundedRectangle(cornerRadius: 16))
                    }

                    // The salon's own opening hours, as the owner sets them in Gestione.
                    salonHours

                    BrandSectionLabel(text: L("profile_account_management"))
                        .padding(.top, 6)
                    StoneKeyValueRow(
                        label: L("profile_change_password"),
                        value: "••••••••",
                        valueColor: .oliveWood,
                        onTap: { changingPassword = true }
                    )
                    .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))

                    BrandSectionLabel(text: L("profile_notifications"))
                        .padding(.top, 6)
                    prefRow(L("profile_pref_reminder"), hint: L("profile_pref_reminder_hint"), keyPath: \.appointmentReminder)
                    prefRow(L("profile_pref_waitlist"), hint: nil, keyPath: \.waitlistAlerts)
                    prefRow(L("profile_pref_marketing"), hint: nil, keyPath: \.marketing)
                    FormErrorBanner(message: viewModel.prefsError)

                    Button {
                        viewModel.logout(onLoggedOut: onLoggedOut)
                    } label: {
                        HStack(spacing: 10) {
                            Image(systemName: "rectangle.portrait.and.arrow.right")
                                .font(.system(size: 15))
                                .foregroundStyle(Color.errorRed)
                            Text(L("profile_logout"))
                                .font(Typo.titleMedium)
                                .foregroundStyle(Color.ink)
                        }
                        .frame(maxWidth: .infinity)
                        .frame(height: 56)
                        .overlay(RoundedRectangle(cornerRadius: 16).strokeBorder(Color.stoneBorder, lineWidth: 1))
                    }
                    .buttonStyle(.plain)
                    .padding(.top, 10)

                    if editing && viewModel.dirty {
                        AccentButton(
                            text: L("profile_save"),
                            action: {
                                viewModel.save()
                                editing = !viewModel.fieldErrors.isEmpty
                            },
                            loading: viewModel.saving,
                            height: 54,
                            corner: 16
                        )
                    }
                }
                .padding(.horizontal, 20)
                .padding(.top, 18)
                .padding(.bottom, 24)
                .readableWidth()
            }
            .onChange(of: viewModel.firstName) { viewModel.fieldChanged() }
            .onChange(of: viewModel.lastName) { viewModel.fieldChanged() }
            .onChange(of: viewModel.email) { viewModel.fieldChanged() }
            .onChange(of: viewModel.phone) { viewModel.fieldChanged() }
            }
        }
        .background(Color.bone)
        .task { await viewModel.load() }
        .sheet(isPresented: $changingPassword) {
            ChangePasswordScreen(
                viewModel: ChangePasswordViewModel(auth: container.auth),
                onDismiss: { changingPassword = false }
            )
        }
    }

    @ViewBuilder
    private var salonHours: some View {
        let groups = viewModel.salon.weeklyHours.groupConsecutiveDays()
        if groups.contains(where: { $0.range != nil }) {
            BrandSectionLabel(text: L("profile_salon_hours"))
                .padding(.top, 6)
            VStack(spacing: 8) {
                ForEach(Array(groups.enumerated()), id: \.offset) { _, group in
                    let closed = group.range == nil
                    StoneKeyValueRow(
                        label: formatDayGroup(group.days),
                        value: group.range.map { "\(formatTime($0.start)) — \(formatTime($0.end))" }
                            ?? L("profile_salon_closed"),
                        valueColor: closed ? .textMuted : .ink
                    )
                    .background(
                        closed
                            ? AnyView(RoundedRectangle(cornerRadius: 10).strokeBorder(Color.stoneBorder, lineWidth: 1))
                            : AnyView(RoundedRectangle(cornerRadius: 10).fill(Color.stone))
                    )
                }
            }
            Text(viewModel.salon.address)
                .font(Typo.jost(12))
                .foregroundStyle(Color.textMuted)
                .padding(.leading, 4)
                .padding(.top, 2)
        }
    }

    private func prefRow(_ title: String, hint: String?, keyPath: WritableKeyPath<ClientNotificationPrefs, Bool>) -> some View {
        HStack {
            VStack(alignment: .leading, spacing: 0) {
                Text(title).font(Typo.bodyLarge).foregroundStyle(Color.ink)
                if let hint {
                    Text(hint).font(Typo.jost(12)).foregroundStyle(Color.textMuted)
                }
            }
            Spacer()
            BrandSwitch(isOn: Binding(
                get: { viewModel.prefs[keyPath: keyPath] },
                set: { newValue in
                    var prefs = viewModel.prefs
                    prefs[keyPath: keyPath] = newValue
                    viewModel.setPrefs(prefs)
                }
            ))
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
        .background(RoundedRectangle(cornerRadius: 16).fill(Color.stone))
    }
}
